File: go/kryon/keyboard_policy_test.go
package kryon

import "testing"

func TestCanonicalKeyboardPolicy(t *testing.T) {
	for mask := 0; mask < 8; mask++ {
		focused, captured, escape := mask&1 != 0, mask&2 != 0, mask&4 != 0
		if Menu_MenuEscapeShouldClose(focused, captured, escape) != (focused && !captured && escape) {
			t.Fatalf("menu focus/capture/Escape mask %d", mask)
		}
		if Menu_MenuContextShouldSuppressClose(focused, captured, escape) != (focused || captured && escape) {
			t.Fatalf("menu opened/trigger/release mask %d", mask)
		}
	}
	for mask := 0; mask < 16; mask++ {
		down, up, right, left := mask&1 != 0, mask&2 != 0, mask&4 != 0, mask&8 != 0
		want := Collapsible_CollapsibleKeyNone()
		switch {
		case down:
			want = Collapsible_CollapsibleKeyDown()
		case up:
			want = Collapsible_CollapsibleKeyUp()
		case right:
			want = Collapsible_CollapsibleKeyRight()
		case left:
			want = Collapsible_CollapsibleKeyLeft()
		}
		if Collapsible_CollapsibleKeyFor(down, up, right, left) != want {
			t.Fatalf("collapsible arrow priority mask %d", mask)
		}
		for _, modifier := range []bool{false, true} {
			got := TextInput_TextShortcutInputFor(modifier, down, up, right, left)
			if got.SelectAll != (modifier && down) || got.Copy != (modifier && up) ||
				got.Cut != (modifier && right) || got.Paste != (modifier && left) {
				t.Fatalf("text shortcut mask %d, modifier %t: %+v", mask, modifier, got)
			}
		}
	}
}

File: tests/keyboard_policy_test.mjs
import assert from "node:assert/strict";
import * as menu from "./menu.js";
import * as collapsible from "./collapsible.js";
import * as text from "./text_input.js";

// Exercise generated decisions, including captured focus and simultaneous keys.
for (const focused of [false, true]) {
  for (const captured of [false, true]) {
    for (const escape of [false, true]) {
      assert.equal(menu.Menu_MenuEscapeShouldClose(null, undefined, undefined,
        focused, captured, escape), focused && !captured && escape);
    }
  }
}
for (const opened of [false, true]) {
  for (const inside of [false, true]) {
    for (const released of [false, true]) {
      assert.equal(menu.Menu_MenuContextShouldSuppressClose(null, undefined, undefined,
        opened, inside, released), opened || (inside && released));
    }
  }
}
for (let mask = 0; mask < 16; mask++) {
  const down = Boolean(mask & 1);
  const up = Boolean(mask & 2);
  const right = Boolean(mask & 4);
  const left = Boolean(mask & 8);
  const expected = down ? collapsible.Collapsible_CollapsibleKeyDown()
    : up ? collapsible.Collapsible_CollapsibleKeyUp()
    : right ? collapsible.Collapsible_CollapsibleKeyRight()
    : left ? collapsible.Collapsible_CollapsibleKeyLeft()
    : collapsible.Collapsible_CollapsibleKeyNone();
  assert.equal(collapsible.Collapsible_CollapsibleKeyFor(null, undefined, undefined,
    down, up, right, left), expected);
  for (const modifier of [false, true]) {
    const shortcuts = text.TextInput_TextShortcutInputFor(null, undefined, undefined,
      modifier, down, up, right, left);
    assert.equal(shortcuts.select_all, modifier && down);
    assert.equal(shortcuts.copy, modifier && up);
    assert.equal(shortcuts.cut, modifier && right);
    assert.equal(shortcuts.paste, modifier && left);
  }
}
console.log("generated JavaScript keyboard policy passed");

for (const [previous, current, boundary] of [
  ["a", ".", true], [".", "b", true], [".", "!", false],
  ["a", " ", false], ["　", "界", true], ["界", "。", true],
  ["界", "β", false],
]) {
  assert.equal(text.TextInput_TextWordBoundaryFor(null, undefined, undefined,
    previous.codePointAt(0), current.codePointAt(0)), boundary);
}
